File: src/net/liplum/blocks/prism/Crystal.java
package net.liplum.blocks.prism;

import arc.graphics.g2d.TextureRegion;
import arc.util.io.Reads;
import arc.util.io.Writes;
import net.liplum.ClientOnly;
import net.liplum.math.PolarPos;
import net.liplum.persistance.RWU;
import net.liplum.utils.ByteU;

public class Crystal {
    public PolarPos revolution;
    public PolarPos rotation;
    @ClientOnly
    public TextureRegion img;
    private static final int ClockwisePos = 0;
    private static final int RemovedPos = 1;
    private static final int AwaitAddingPos = 2;
    private static final int Pos4 = 3;
    private static final int Pos5 = 4;
    private static final int Pos6 = 5;
    private static final int Pos7 = 6;
    private static final int Pos8 = 7;

    public int orbitPos = 0;
    public int data = 0;

    public boolean isClockwise() {
        return ByteU.isOn(data, ClockwisePos);
    }

    public void setClockwise(boolean clockwise) {
        if (clockwise)
            data = ByteU.on(data, ClockwisePos);
        else
            data = ByteU.off(data, ClockwisePos);
    }

    public boolean isRemoved() {
        return ByteU.isOn(data, RemovedPos);
    }

    public void setRemoved(boolean removed) {
        if (removed) {
            data = ByteU.on(data, RemovedPos);
            setAwaitAdding(false);
        } else
            data = ByteU.off(data, RemovedPos);
    }

    public boolean isAwaitAdding() {
        return ByteU.isOn(data, AwaitAddingPos);
    }

    public void setAwaitAdding(boolean awaitAdding) {
        if (awaitAdding) {
            data = ByteU.on(data, AwaitAddingPos);
            setRemoved(false);
        } else
            data = ByteU.off(data, AwaitAddingPos);
    }

    public static void write(Writes writes, Crystal crystal) {
        writes.b(crystal.orbitPos);
        RWU.writePolarPos(writes, crystal.revolution);
        RWU.writePolarPos(writes, crystal.rotation);
        writes.b(crystal.data);
    }

    public static Crystal read(Reads reads) {
        Crystal crystal = new Crystal();
        crystal.orbitPos = reads.b();
        crystal.revolution = RWU.readPolarPos(reads);
        crystal.rotation = RWU.readPolarPos(reads);
        crystal.data = reads.b();
        return crystal;
    }
}

File: main/src/net/liplum/blocks/prism/Crystal.kt
package net.liplum.blocks.prism

import arc.util.io.Reads
import arc.util.io.Writes
import net.liplum.Var
import net.liplum.common.math.PolarX
import net.liplum.common.persistence.CacheReaderSpec
import net.liplum.common.persistence.CacheWriter
import net.liplum.common.persistence.IRWableX
import net.liplum.common.util.isOn
import net.liplum.common.util.off
import net.liplum.common.util.on
import plumy.core.ClientOnly
import plumy.core.assets.TR
import java.io.DataInputStream

class Crystal : IRWableX {
    var revolution = PolarX()
    var rotation = PolarX()
    @ClientOnly
    var img = TR()
    @ClientOnly
    var lastPassThroughTime = Var.PrismCrystalPassThroughBloomTime
    val isActivated get() = lastPassThroughTime < Var.PrismCrystalPassThroughBloomTime
    var orbitPos = 0
    var data = 0
    var isClockwise: Boolean
        get() = data.isOn(ClockwisePos)
        set(clockwise) {
            data = if (clockwise) data.on(ClockwisePos) else data.off(ClockwisePos)
        }
    var isRemoved: Boolean
        get() = data.isOn(RemovedPos)
        set(removed) {
            if (removed) {
                data = data.on(RemovedPos)
                isAwaitAdding = false
            } else data = data.off(RemovedPos)
        }
    var isAwaitAdding: Boolean
        get() = data.isOn(AwaitAddingPos)
        set(awaitAdding) {
            if (awaitAdding) {
                data = data.on(AwaitAddingPos)
                isRemoved = false
            } else data = data.off(AwaitAddingPos)
        }

    companion object {
        private const val ClockwisePos = 0
        private const val RemovedPos = 1
        private const val AwaitAddingPos = 2
        private const val Pos4 = 3
        private const val Pos5 = 4
        private const val Pos6 = 5
        private const val Pos7 = 6
        private const val Pos8 = 7
    }

    override fun read(reader: Reads) = reader.run {
        orbitPos = b().toInt()
        revolution.read(this)
        rotation.read(this)
        data = b().toInt()
    }

    override fun read(reader: DataInputStream) = CacheReaderSpec(reader).run {
        orbitPos = b().toInt()
        revolution.read(reader)
        rotation.read(reader)
        data = b().toInt()
    }

    override fun write(writer: Writes) = writer.run {
        b(orbitPos)
        revolution.write(this)
        rotation.write(this)
        b(data)
    }

    override fun write(writer: CacheWriter) = writer.run {
        b(orbitPos)
        revolution.write(this)
        rotation.write(this)
        b(data)
    }
}
